conda/core/solve.py

```diff
@@ -31,18 +38,341 @@
+
+        solution = IndexedSet(r.dependency_sort({d.name: d for d in solution}))
+        log.debug("solved prefix %s\n"
+                  "  solved_linked_dists:\n"
+                  "    %s\n",
+                  self.prefix, "\n    ".join(text_type(d) for d in solution))
+        solution = IndexedSet(index[d] for d in solution)
+
+        return solution
+
+
+
+    def solve_for_diff(self, prune=False, force_reinstall=False, deps_modifier=None,
+                          ignore_pinned=False):
+        """Gives the package references to remove from an environment, followed by
+        the package references to add to an environment.
+        
+        Args:
+            prune (bool):
+                See :meth:`solve_final_state`.
+            force_reinstall (bool):
+                See :meth:`solve_final_state`.
+            deps_modifier (DepsModifier):
+                See :meth:`solve_final_state`.
+            ignore_pinned (bool):
+                See :meth:`solve_final_state`.
+
+        Returns:
+            Tuple[PackageRef], Tuple[PackageRef]:
+
+        """
+        # need to take the final state, but force_reinstall and deps_modifier are
+        # also relevant here (also add back requested specs if force_reinstall)
+        force_reinstall = force_reinstall is NULL and context.force or force_reinstall
+        final_dists = self.solve_final_state(prune, force_reinstall, deps_modifier, ignore_pinned)
+
+        previous_dists = IndexedSet(iterkeys(linked_data(self.prefix)))
+
+        unlink_dists = previous_dists - final_dists
+        link_dists = final_dists - previous_dists
+
+        # TODO: add back 'noarch: python' to unlink and link if python version changes
+
+        return unlink_dists, link_dists
+
+
+    def solve_for_transaction(self, prune=False, force_reinstall=False, deps_modifier=None,
+                          ignore_pinned=False):
+        """
+        
+        Args:
+            prune (bool):
+                See :meth:`solve_final_state`.
+            force_reinstall (bool):
+                See :meth:`solve_final_state`.
+            deps_modifier (DepsModifier):
+                See :meth:`solve_final_state`.
+            ignore_pinned (bool):
+                See :meth:`solve_final_state`.
+
+        Returns:
+            UnlinkLinkTransaction:
+
+        """
+        if self.prefix == context.root_prefix and context.enable_private_envs:
+            # hold on, it's a while ride
+            pass
+        else:
+            unlink_dists, link_dists = self.solve_for_diff(prune, force_reinstall,
+                                                           deps_modifier, ignore_pinned)
+            unlink_dists = tuple(Dist(d) for d in unlink_dists)
+            link_dists = tuple(Dist(d) for d in link_dists)
+            stp = PrefixSetup(self.index, self.prefix, unlink_dists, link_dists, 'INSTALL',
+                              self.specs_to_add)
+            return UnlinkLinkTransaction(stp)
```

Comment:
This separation into the three distinct `solve_...` methods looks great ! Why not go a step further and remove the second (`solve_for_diff`) and third (`solve_for_transaction`) from the `Solver` class and make them simple functions whose input is the output of `Solver.solve_final_state` ? It seems to me that would simplify the API as then the Solver wouldn't have to know anything about files, links, and transactions. It would simply be a computation device that transforms certain models (such as `MatchSpec` and `PackageRecord`) into other models (`PackageRef`).